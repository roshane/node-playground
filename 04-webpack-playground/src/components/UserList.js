import { useEffect, useState } from 'react';

const url = 'https://jsonplaceholder.typicode.com/users'

const UserList = () => {
    const [userList, setUserList] = useState([]);
    const [isLoading, setLoading] = useState(false);
    const [loadUserTimestamp, setLoadUserTimestamp] = useState(0);

    useEffect(() => {
        setLoading(true);
        fetch(url)
            .then(response => response.json())
            .then(json => setUserList(json))
            .then(() => setLoading(false));
    }, [loadUserTimestamp]);


    return (
        <>
            <button
                className='rounded bg-orange-300 px-2 py-2 text-black hover:bg-orange-400'
                onClick={() => setLoadUserTimestamp(new Date().getTime())}
                data-testid='btnRefresh'>
                Refresh
            </button>
            {
                isLoading ? <p data-testid='loadingMsg'>Loading...</p> : <div data-testid='userListView'>
                    {
                        userList.map(user => (
                            <div key={user.id}>
                                <pre>
                                    {JSON.stringify(user, null, 2)}
                                </pre>
                                <hr />
                            </div>
                        ))
                    }
                </div>
            }
        </>
    );

}

export default UserList;