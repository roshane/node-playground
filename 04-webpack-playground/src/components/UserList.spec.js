import '@testing-library/jest-dom';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import UserList from './UserList';

describe("<UserList/>", () => {

    let mockFetch;
    beforeEach(() => {
        mockFetch = jest.fn();
        global.fetch = mockFetch;
    });

    afterEach(()=>{
        cleanup();
        mockFetch.mockReset();
    });

    test('<UserList/> should be emtpy initially', () => {
        const component = TestRenderer.create(<UserList />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    const fakeUserListJson = [{
        "id": 1,
        "name": "Leanne Graham",
        "username": "Bret",
        "email": "Sincere@april.biz",
        "address": {
            "street": "Kulas Light",
            "suite": "Apt. 556",
            "city": "Gwenborough",
            "zipcode": "92998-3874",
            "geo": {
                "lat": "-37.3159",
                "lng": "81.1496"
            }
        },
        "phone": "1-770-736-8031 x56442",
        "website": "hildegard.org",
        "company": {
            "name": "Romaguera-Crona",
            "catchPhrase": "Multi-layered client-server neural-net",
            "bs": "harness real-time e-markets"
        }
    }];

    test('<UserList/> should load users upon [Refresh] button click', async () => {
        mockFetch.mockReturnValue(Promise.resolve({
            json: () => Promise.resolve(fakeUserListJson)
        }));
        const { getByTestId } = render(<UserList />);
        fireEvent.click(getByTestId('btnRefresh'));
        await waitFor(() => {
            expect(getByTestId('userListView')).toBeInTheDocument()
        });
    });
});