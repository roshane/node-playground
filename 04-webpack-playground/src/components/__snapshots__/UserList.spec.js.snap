// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList/> <UserList/> should be emtpy initially 1`] = `
Array [
  <button
    data-testid="btnRefresh"
    onClick={[Function]}
  >
    Refresh
  </button>,
  <div
    data-testid="userListView"
  />,
]
`;
